### ISSUE TITLE
[Bug]: all references to metadata(in an Item element inside a Target) always refer to the value prior to evaluating the Item element

### ISSUE THREAD

**Issue by Arthri** [dotnet/msbuild#10721]
### Issue Description

When using item metadata within an Item element contained inside a Target, the value of the metadata always evaluates to its value immediately before entering the Item element. That is, any updates to metadata inside an Item element are not reflected in later references to that metadata inside the same Item element

### Steps to Reproduce

A minimal sample project is available at https://github.com/Arthri/msbuild_demo_batching

1. Include some items into a group
2. Create a target which operates over the items and changes some metadata
3. In the same Item element, update metadata conditionally using the metadata that was changed

Miminal sample
```xml
<ItemGroup>
  <T Include="a;b;c;d" />
</ItemGroup>

<Target Name="_T" BeforeTargets="Build">
  <ItemGroup>
    <T>
      <M>default data</M>
      <M Condition="'%(T.M)' == 'default data'">new default</M>
    </T>
  </ItemGroup>

  <!-- Outputs 'default data' four times -->
  <Warning Text="%(T.Identity): %(T.M)" />
</Target>
```

### Expected Behavior

The condition should've used the new value of the metadata, and outputted four warnings containing "new default"

### Actual Behavior

The condition used the old value of the metadata: the metadata before entering the Item element in the ItemGroup, and outputted four warnings containing "default data"

### Analysis

I believe that MSBuild is replacing the metadata before entering the Item element, and when the condition runs, it never sees the updated metadata values.

Assuming the aforementioned is true, it can be concluded that spreading the metadata updates over several Item elements should work. And from my testing, I've observed that this is a working solution
```xml
<ItemGroup>
  <T>
    <M>default data</M>
  </T>
  <T>
    <M Condition="'%(T.M)' == 'default data'">new default</M>
  </T>
</ItemGroup>
```

### Versions & Configurations

.NET SDK version `8.0.400`
MSBuild version `17.11.3.35201` / `17.11.3+0c8610977`
Microsoft Visual Studio version `17.11.3`

**Comment by AR-May** [comment]
@rainersigwald It seems like [documentation](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-conditions?view=vs-2022#metadata) concerning expansion in conditions does not mention this situation in detail: "The value of an %-expression in a target is evaluated at run-time and depends on any state changes during target execution.".

**Comment by rainersigwald** [comment]
Yes we should mention this behavior in the docs, it's not the behavior I'd expect but I don't expect it to change in the future either.
